### PR TITLE
refactor(@angular-devkit/build-angular): simplify how specs are found and loaded

### LINKED ISSUE #22531: Running tests with code coverage doesn't include files outside the sourceRoot directory
# 🐞 Bug report

### Command (mark with an `x`)

- [ ] new
- [ ] build
- [ ] serve
- [x] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] extract-i18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc

### Is this a regression?

 Yes, the previous version in which this bug was not present was: Any time before PR #11974

### Description

 A clear and concise description of the problem...

## 🔬 Minimal Reproduction

1. ng new test-app
2. create packages/test-package outside of src folder
3. add there simple test.component.ts and test.component.spec.ts
4. Modify test.ts file and add context of 'packages directory
replace
```javascript
const context = require.context('./', true, /\.spec\.ts$/);
context.keys().map(context1);
```
with
```javascript
const context1 = require.context('../packages', true, /\.spec\.ts$/);
const context2 = require.context('./', true, /\.spec\.ts$/);

context1.keys().map(context1);
context2.keys().map(context2);
```
5. Add "../packages/**/*.spec.ts", to tsconfig.spec.json in "include" section

6. Run `ng test --code-coverage` and open coverage/index.html

## 🔥 Exception or Error

No error but the files outside the directory are not included in the coverage report.

**Anything else relevant?**

This was reported previously on issue #12474. They simply dismissed it as saying only files under `sourceRoot` are listed on the coverage report. Having the intended behavior be where tests from outside `sourceRoot` are ran but not included in the coverage report seems counterintuitive. They should either both be ran and included in the coverage report or not included in either place.

### LINKED ISSUE #23751: ng test exits with success even if test code does not compile
# 🐞 Bug report

### Command (mark with an `x`)

- [ ] new
- [ ] build
- [ ] serve
- [x] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] extract-i18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc

### Is this a regression?

 Yes, the previous version in which this bug was not present was: 13.2.2

### Description

A regression introduced in v13.2.3 of the CLI has been allowing test code that doesn't even compile to appear as a passing job in our CICD system.

The problem can be traced to [this change](https://github.com/angular/angular-cli/pull/22670/files#r801922504) which includes an unconditional promise resolve following compilation, no matter if there are compilation errors. I think I understand the intent here, but I believe the `singleRun` scenario was overlooked. When running in watch mode, it's true that you always want Karma to start up and the exit code of the command is irrelevant, since the command runs until the user forces it to stop. But within CICD systems `singleRun` is the norm and the exit code is very important. We do not want karma to start if the compilation fails and the command should exit with a non zero code.

I think all that is required is to add a reject handler and call that one instead it there are errors `&& singleRun`. I have verified this fix locally and will open a PR with precisely this change against this issue.

## 🔬 Minimal Reproduction

Create an Angular project using ng cli 13.2.3 or above.

Create one unit test. Include a compilation error in the test.

Ensure the karma option `failOnEmptyTestSuite` is `false`.

Run `ng test --project=<project-name> --watch=false`

Observe that karma starts despite the compilation error and the command exits with success (0 code).

This is all captured in the following reproduction repo: https://github.com/vigie/false-positive-repo-app

## 🔥 Exception or Error

<pre><code>

</code></pre>

## 🌍 Your Environment

<pre><code>
     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/

Angular CLI: 13.3.8
Node: 14.20.0
Package Manager: npm 6.14.17
OS: linux x64

Angular: 13.3.11
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... platform-server, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1303.8
@angular-devkit/build-angular   13.3.8
@angular-devkit/core            13.3.8
@angular-devkit/schematics      13.3.8
@angular/cdk                    13.3.9
@angular/cli                    13.3.8
@schematics/angular             13.3.8
ng-packagr                      13.3.1
rxjs                            6.6.7
typescript                      4.6.3
webpack                         5.73.0
</code></pre>

**Anything else relevant?**

As mentioned, this was introduced at v13.2.3 and as far as I can tell still exists in latest v14. The fix should be backported and released in v13 also.

### PR DESCRIPTION
**Pull request by alan-agius4**
Prior to this change specs where found and loaded using Webpack's `require.context` API. The `require.context` is found in the users project `test.ts`. This resulted in a complex and hacky setup especially to filter tests when the `include` builder option is provided, were we had to amend the `test.ts` in memory.

With this change we find all the specs files and add them as part of the main entrypoint.

Closes #23751 and closes #22531

### CODE REVIEW

File: packages/angular_devkit/build_angular/src/builders/karma/index.ts [hunk 1]
```diff
@@ -110,45 +106,22 @@ export function execute(
...
-            loader: SingleTestTransformLoader,
-            options: {
-              files,
-              logger: context.logger,
-            },
-          },
-        });
-      }
+      webpackConfig.plugins ??= [];
+      webpackConfig.plugins?.push(
```
Comment (clydin) [inline packages/angular_devkit/build_angular/src/builders/karma/index.ts:118]: nit: `plugins` should always be defined via the previous statement

File: packages/angular_devkit/build_angular/src/webpack/plugins/karma/karma.ts [hunk 2]
```diff
@@ -124,10 +124,10 @@ const init: any = (config: any, emitter: any) => {
 
       logger.error(statsErrorsToString(statsJson, { colors: true }));
 
-      // Notify potential listeners of the compile error.
-      emitter.emit('compile_error', {
-        errors: statsJson.errors?.map((e) => e.message),
-      });
+      if (config.singleRun) {
+        // Notify potential listeners of the compile error.
+        emitter.emit('load_error');
```
Comment (clydin) [inline packages/angular_devkit/build_angular/src/webpack/plugins/karma/karma.ts:129]: Why the change to the event name here?

File: packages/angular_devkit/build_angular/src/builders/karma/find-tests-plugin.ts [hunk 3]
```diff
@@ -6,15 +6,76 @@
...
+
+export class FindTestsPlugin {
+  private compilation: Compilation | undefined;
+
+  constructor(private options: FindTestsPluginOptions) {}
+
+  apply(compiler: Compiler): void {
+    const { include = ['**/*.spec.ts'], projectSourceRoot, workspaceRoot } = this.options;
+
+    compiler.hooks.environment.tap(PLUGIN_NAME, () => {
```
Comment (clydin) [inline packages/angular_devkit/build_angular/src/builders/karma/find-tests-plugin.ts:38]: Is this hook needed?  Could the compiler options be modified directly in the apply function?

File: packages/angular_devkit/build_angular/src/webpack/plugins/karma/karma.ts (hunk 2 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/src/webpack/plugins/karma/karma.ts:129]: load_error is needed as otherwise in some cases karma will jot exit when there is a webpack error as karma doesn’t listen to compilation_error.

File: packages/angular_devkit/build_angular/src/builders/karma/find-tests-plugin.ts (hunk 3 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/src/builders/karma/find-tests-plugin.ts:38]: I believe so, will update.

File: packages/angular_devkit/build_angular/src/builders/karma/index.ts (hunk 1 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/src/builders/karma/index.ts:118]: You are right 👍

**Review by clydin — APPROVED** [review]

**Comment by vigie** [comment]
Is it possible to backport this to the 14 branch so we can consume the fix to https://github.com/angular/angular-cli/issues/23751?
